LE Audio: Add Unicast client operation callbacks

Thalley opened the pull request:
Adds unicast client callbacks for each BAP operation, which allows an application to properly handle any rejections/errors returned by the unicast server.

Thalley committed:
Bluetooth: Audio: Fixed naming of invalid length ASCS response

The "Invalid Length" response code was called truncated,
which does not match the spec definition.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add unicast client cp notify callbacks

Add callbacks for control point notifications sent by the
unicast server. These will allow the upper layers to know
if and why a specific operation failed.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Shell: Add unicast client operation callbacks

Add the unicast client operation callbacks to the shell, simply
printing the values sent by the server.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
tests: Bluetooth: Audio: Add unicast client CP notification bsim testing

Adds waiting for the control point operation callbacks in the
unicast client BSIM test, ensuring that the callbacks work, but also
testing that the server is correctly sending them.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

MariuszSkamra commented on include/zephyr/bluetooth/audio/audio.h:
```diff
@@ -1354,6 +1425,116 @@ struct bt_audio_unicast_client_cb {
+	 * @param reason   Reason code.
+	 */
+	void (*config)(struct bt_audio_stream *stream,
+		       enum bt_ascs_rsp_code rsp_code,
+		       enum bt_ascs_reason reason);
```
Did you consider adding dedicated container, like:
```
struct bt_ascs_rsp {
	 enum bt_ascs_rsp_code code;
	 enum bt_ascs_reason reason;
};
```

Thalley replied:
I haven't :) What advantages do you see of that approach?

MariuszSkamra commented on include/zephyr/bluetooth/audio/audio.h:
```diff
@@ -93,6 +93,77 @@ enum bt_audio_parental_rating {
+ * These are sent by the server to the client when a stream operation is
+ * requested.
+ */
+enum bt_ascs_rsp_code {
+	/** Server completed operation successfully */
+	BT_ASCS_RSP_CODE_SUCCESS              = 0x00,
+	/** Server did not support operation by client */
+	BT_ASCS_RSP_CODE_NOT_SUPPORTED        = 0x01,
+	/** Server rejected due to invalid operation length */
+	BT_ASCS_RSP_CODE_INVALID_LENGTH       = 0x02,
```
I am not convinced about the naming. I know that those values come from ASCS specification but in audio.h API we don't use the ASCS prefix

Thalley replied:
Me neither. 

That's also a challenge I had with BASS; each service may define a set of error/response codes, and we can't just prefix them with `bt_bap` or `bt_audio`, as they differ. 

We could of course convert them to e.g. errno values or something as well. 

So basically should we expose service specific error codes in the public API at all? And if so, how should we group/prefix them?

Thalley replied:
> So basically should we expose service specific error codes in the public API at all? And if so, how should we group/prefix them?

We could also have `BT_BAP_ERR` codes, and transform from BASS and ASCS to a public BT_BAP one.

Thalley committed:
Bluetooth: Audio: Add special handling of release cp notification

When releasing a stream as the unicast client, we may get
either the ASE state notification with state idle or
the control point notification first. If we get the
ASE state notification first, we pretend we get the
CP notification first, and call the release callback.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented:
Rebased to solve merge conflicts

Thalley commented:
Rebased to solve merge conflicts

cvinayak reviewed (approved):
QUICK-ACK. LGTM.
